### PR TITLE
Login page

### PR DESCRIPTION
**Pull request by IchabodCrow**

### CODE REVIEW

File: src/App.js [hunk 1]
```diff
@@ -1,26 +1,23 @@
...
-          className="App-link"
-          href="https://reactjs.org"
-          target="_blank"
-          rel="noopener noreferrer"
-        >
-          Learn React
-        </a>
-      </header>
+      <Switch>
+        <Route exact path={"/movie"} component={Main}/>
```
Comment (reshetov) [inline src/App.js]: I can open this route without login. It's bad. Try to use PrivateRoute instead - https://medium.com/@tomlarge/private-routes-with-react-router-dom-28e9f40c7146

File: src/components/Login/LoginForm.jsx [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+  email,
+  requiredEmail,
+  requiredPassword,
+} from "../../utils/validatorsFields";
+import Input from "../UI/Input";
+
+const LoginForm = (props) => {
+  const { handleSubmit } = props;
+  return (
+    <div className="flex justify-center mr-8">
```
Comment (reshetov) [inline src/components/Login/LoginForm.jsx]: Install prettier (https://prettier.io/) in your editor and use it

File: src/components/Login/LoginForm.jsx [hunk 3]
```diff
@@ -0,0 +1,108 @@
+import React from "react";
+import { reduxForm, Field, SubmissionError } from "redux-form";
```
Comment (reshetov) [inline src/components/Login/LoginForm.jsx]: Please use Final-form instead Redux-form (https://final-form.org/react)

File: src/components/Login/LoginForm.jsx [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+import { withRouter, useHistory } from "react-router-dom";
+import {
+  email,
+  requiredEmail,
+  requiredPassword,
+} from "../../utils/validatorsFields";
+import Input from "../UI/Input";
+
+const LoginForm = (props) => {
```
Comment (reshetov) [inline src/components/Login/LoginForm.jsx:9]: Split with MEGA-file to separated components

File: src/components/Login/LoginForm.jsx [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+  return (
+    <div className="flex justify-center mr-8">
+      <form onSubmit={handleSubmit} className="w-full max-w-sm">
+        <div className="md:flex md:items-center mb-6">
+          <div className="md:w-1/3 mb-5">
+            <label
+              className="block text-gray-500 font-bold md:text-right mb-1 md:mb-0 pr-4"
+              htmlFor="inline-full-name"
+            >
+              Email
```
Comment (reshetov) [inline src/components/Login/LoginForm.jsx]: Use locales for text - https://github.com/i18next/react-i18next

File: src/utils/validatorsFields.js [hunk 6]
```diff
@@ -0,0 +1,7 @@
+export const email = value =>
```
Comment (reshetov) [inline src/utils/validatorsFields.js]: Rename to `emailRegex`

File: src/utils/validatorsFields.js [hunk 7]
```diff
@@ -0,0 +1,7 @@
...
+  value && !/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i.test(value) ?
+  'Не корректный email' : undefined
```
Comment (reshetov) [inline src/utils/validatorsFields.js]: Locale

File: tailwind.js [hunk 8]
```diff
@@ -0,0 +1,774 @@
+module.exports = {
```
Comment (reshetov) [inline tailwind.js:1]: Remove this file. It's redundant

File: src/components/Login/LoginForm.jsx [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+      </form>
+    </div>
+  );
+};
+
+const LoginReduxForm = reduxForm({
+  form: "login",
+})(LoginForm);
+
+const Login = () => {
```
Comment (reshetov) [inline src/components/Login/LoginForm.jsx]: Move to component

File: src/components/Login/LoginForm.jsx [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+  const history = useHistory();
+  const registerdUser = JSON.parse(localStorage.getItem("user"));
```
Comment (reshetov) [inline src/components/Login/LoginForm.jsx]: Typo. `registered`

Comment (reshetov) [reply src/components/Login/LoginForm.jsx]: What you will do if you have more that one user in DB? Perhaps be better to implement `checkUserCredentials` method in some service which will accepts login/password args and checks that they have in DB (LS) and valid?

File: src/components/Login/LoginForm.jsx [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+
+  const handleSubmit = (user) => {
+    const sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms));
```
Comment (reshetov) [inline src/components/Login/LoginForm.jsx]: WAT?

File: src/components/Login/LoginForm.jsx [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+    return sleep(1000).then(() => {
+      if (
+        registerdUser.email === user.email &&
+        registerdUser.password === user.password
+      ) {
+        history.push("/movie");
+      } else {
+        throw new SubmissionError({
+          email: "Упс. Где-то ошибка",
```
Comment (reshetov) [inline src/components/Login/LoginForm.jsx]: Just display error label near login form

File: src/components/Login/LoginForm.jsx [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+        });
+      }
+    });
+  };
+
+  return (
+    <div className="flex flex-col">
+      <div className="my-8 w-1/3 self-center">
+        <h1 className="font-bold text-4xl bg-purple-500 text-center text-white font-bold py-2 px-4 rounded ">
+          My favorite cinema
```
Comment (reshetov) [inline src/components/Login/LoginForm.jsx]: locale

File: src/components/UI/Input.js [hunk 14]
```diff
@@ -0,0 +1,25 @@
+import React from "react";
+
+const Input = ({
+  placeholder,
+  input,
+  type,
+  meta: { touched, error, warning },
```
Comment (reshetov) [inline src/components/UI/Input.js]: Where are you get this? This code doesn't used

File: src/containers/LoginContainers.js [hunk 15]
```diff
@@ -0,0 +1,22 @@
+import { connect } from 'react-redux'
```
Comment (reshetov) [inline src/containers/LoginContainers.js]: What does it do?

File: src/services/insteadServer.js [hunk 16]
```diff
@@ -0,0 +1,13 @@
+const InsteadServer = function () {
+  this.prepareInitialData = () => {
+    const user = { email : "fof@fof.ru", password: "123"}
```
Comment (reshetov) [inline src/services/insteadServer.js]: Add more that one user

File: src/store/store.js [hunk 17]
```diff
@@ -0,0 +1,18 @@
+import { createStore, compose, applyMiddleware, combineReducers } from "redux";
+import { reducer as formReducer } from 'redux-form'
+import thunk from 'redux-thunk'
+
+const middelewares = [thunk]
+
+const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION__ ? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({
```
Comment (reshetov) [inline src/store/store.js:7]: Why is it?

Comment (IchabodCrow) [reply src/store/store.js:7]: To use Redux dev tools in a browser

File: src/components/Login/LoginForm.jsx [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+              type="password"
+              placeholder={i18n.t("description.passwordPlaceholder")}
+              labelText={i18n.t("description.password")}
+              validate={composeValidators(required)}
+            />
+            <div className="md:flex md:items-center">
+              <div className="md:w-1/3"></div>
+              <div className="md:w-2/3">
+                <button className="shadow bg-purple-500 hover:bg-purple-400 focus:shadow-outline focus:outline-none text-white font-bold py-2 px-4 rounded">
+                  <Trans i18nKey="description.signIn">
```
Comment (k41n) [inline src/components/Login/LoginForm.jsx]: Why not just `{t('description.signIn')}`? Why do we need `<Trans>`?

File: src/components/Login/LoginForm.jsx [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+const LoginForm = (props) => {
+  return (
+    <Form
+      onSubmit={props.handleSubmit}
+      render={({ submitError, handleSubmit }) => (
+        <div className="flex justify-center mr-8">
+          <form onSubmit={handleSubmit} className="w-full max-w-sm">
+            <FieldWithLabel
+              name="email"
+              type="text"
```
Comment (k41n) [inline src/components/Login/LoginForm.jsx]: For email input there is pretty good `email` type

File: src/components/Login/LoginForm.jsx [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+              validate={composeValidators(required, emailRegex)}
+            />
+            <FieldWithLabel
+              name="password"
+              type="password"
+              placeholder={i18n.t("description.passwordPlaceholder")}
+              labelText={i18n.t("description.password")}
+              validate={composeValidators(required)}
+            />
+            <div className="md:flex md:items-center">
```
Comment (k41n) [inline src/components/Login/LoginForm.jsx]: All this looks like pretty good `<Actions>` component we probably will be able to reuse in other form, won't we?

File: src/components/Login/LoginForm.jsx [hunk 21]
```diff
@@ -0,0 +1,52 @@
...
+import FieldWithLabel from "../UI/FieldWithLabel";
+
+const LoginForm = (props) => {
+  return (
+    <Form
+      onSubmit={props.handleSubmit}
+      render={({ submitError, handleSubmit }) => (
+        <div className="flex justify-center mr-8">
+          <form onSubmit={handleSubmit} className="w-full max-w-sm">
+            <FieldWithLabel
```
Comment (k41n) [inline src/components/Login/LoginForm.jsx]: Looks like `<EmailField name="email" namespace="description" required>`

File: src/components/Login/LoginForm.jsx [hunk 22]
```diff
@@ -0,0 +1,52 @@
...
+              name="email"
+              type="text"
+              placeholder={i18n.t("description.emailPlaceholder")}
+              labelText={i18n.t("description.email")}
+              validate={composeValidators(required, emailRegex)}
+            />
+            <FieldWithLabel
```
Comment (k41n) [inline src/components/Login/LoginForm.jsx]: Looks like `<PasswordField name="password" namespace="description" required />`

File: src/components/Login/LoginPage.jsx [hunk 23]
```diff
@@ -0,0 +1,39 @@
...
+      return { [FORM_ERROR]: `${i18n.t("validation.submitError")}` }
+    }
+    
+  };
+
+  return (
+    <div className="flex flex-col">
+      <div className="my-8 w-1/3 self-center">
+        <h1 className="font-bold text-4xl bg-purple-500 text-center text-white font-bold py-2 px-4 rounded ">
+          <Trans i18nKey="title">
```
Comment (k41n) [inline src/components/Login/LoginPage.jsx]: Why `<Trans />`? We can get `t` with `useTranslation` and use it

File: src/components/Login/LoginPage.jsx [hunk 24]
```diff
@@ -0,0 +1,39 @@
...
+    } else {
+      return { [FORM_ERROR]: `${i18n.t("validation.submitError")}` }
+    }
+    
+  };
+
+  return (
+    <div className="flex flex-col">
+      <div className="my-8 w-1/3 self-center">
+        <h1 className="font-bold text-4xl bg-purple-500 text-center text-white font-bold py-2 px-4 rounded ">
```
Comment (k41n) [inline src/components/Login/LoginPage.jsx]: Looks like we can extract that to `<PageHeader title="title" />` component

File: src/components/Login/LoginPage.jsx [hunk 25]
```diff
@@ -0,0 +1,39 @@
+import React from "react";
+import { withRouter, useHistory } from "react-router-dom";
+import { Trans, useTranslation } from 'react-i18next'
+import { FORM_ERROR } from 'final-form'
+
+import i18n from '../../i18n'
+import LoginForm from "./LoginForm";
+import checkUserCredentails from "../../services/checkUserCredentails";
```
Comment (k41n) [inline src/components/Login/LoginPage.jsx]: Unused?

File: src/components/UI/FieldWithLabel.jsx [hunk 26]
```diff
@@ -0,0 +1,45 @@
+import React from "react";
+import { Field } from "react-final-form";
+import { useTranslation } from 'react-i18next'
```
Comment (k41n) [inline src/components/UI/FieldWithLabel.jsx]: Unused?

File: src/components/UI/FieldWithLabel.jsx [hunk 27]
```diff
@@ -0,0 +1,45 @@
...
+      <div className="md:w-1/3 mb-5">
+        <label
+          className="block text-gray-500 font-bold md:text-right mb-1 md:mb-0 pr-4"
+          htmlFor="inline-full-name"
+        > 
+            {labelText}
+        </label>
+      </div>
+      <div className="md:w-2/3 h-16">
+        <Field
```
Comment (k41n) [inline src/components/UI/FieldWithLabel.jsx]: Ok, maybe specify subscription? Are we using all the fields of the FieldState? Are we going to rerender on every change? Subscribe only to what really matters.

File: src/components/UI/FieldWithLabel.jsx [hunk 28]
```diff
@@ -0,0 +1,45 @@
...
+          {({ input, meta }) => (
+            <div>
+              <input
+                {...input}
+                type="text"
+                placeholder={placeholder}
+                className="bg-gray-200 appearance-none border-2 border-gray-200 rounded w-full py-2 px-4 text-gray-700 leading-tight focus:outline-none focus:bg-white focus:border-purple-500"
+                id="inline-full-name"
+              />
+              {meta.error && meta.touched && (
```
Comment (k41n) [inline src/components/UI/FieldWithLabel.jsx]: Hide error if field is `modifiedSinceLastSubmit` - so when user starts to fix, error is hidden

File: src/App.js [hunk 29]
```diff
@@ -1,26 +1,34 @@
...
-        <a
-          className="App-link"
-          href="https://reactjs.org"
-          target="_blank"
-          rel="noopener noreferrer"
-        >
-          Learn React
-        </a>
-      </header>
+      <button onClick={ () => changeLanguage('ru') }>Русский</button>
```
Comment (k41n) [inline src/App.js]: Вёрстка этого куска просто божественна
image

File: src/index.js [hunk 30]
```diff
@@ -1,17 +1,25 @@
...
-import * as serviceWorker from './serviceWorker';
+import React, { Suspense }  from "react"
+import ReactDOM from "react-dom"
+import { Provider } from "react-redux"
+import { BrowserRouter as Router } from "react-router-dom"
+
+import './i18n';
+import App from "./App"
+import * as serviceWorker from "./serviceWorker"
+import store from "./store/store"
```
Comment (k41n) [inline src/index.js:9]: А зачем тут вообще нужен Redux?

File: src/services/checkUserCredentails.js [hunk 31]
```diff
@@ -0,0 +1,16 @@
+const CheckUserCredentails = function () {
+  this.checkUserRegistred = (value) => {
```
Comment (k41n) [inline src/services/checkUserCredentails.js]: Typo

File: src/services/checkUserCredentails.js [hunk 32]
```diff
@@ -0,0 +1,16 @@
...
+    const findUser = users.findIndex(function (item) {
+      return item.email === value.email && item.password === value.password
+        ? true
+        : false;
+    });
+    return findUser;
+  };
+};
+
+const checkUserCredentails = new CheckUserCredentails();
```
Comment (k41n) [inline src/services/checkUserCredentails.js]: Зачем тут класс вообще? Почему бы не экспортировать просто функцию?

File: src/services/insteadServer.js [hunk 33]
```diff
@@ -0,0 +1,26 @@
...
+  this.setToken = (user) => {
+    const checkUserRegister = checkUserCredentails.checkUserRegistred(user);
+    if (checkUserRegister !== -1) {
+      const authToken = "hi_i_am_a_temporary_token_for_setting_up_routing";
+      localStorage.setItem("token", authToken);
+    }
+  };
+};
+
+const insteadServer = new InsteadServer();
```
Comment (k41n) [inline src/services/insteadServer.js]: Зачем тут класс? Почему бы просто не экспортировать две функции?

File: src/store/store.js [hunk 34]
```diff
@@ -0,0 +1,17 @@
...
+trace: true,
+raceLimit: 55,
+}) : compose;
+
+const reducers = combineReducers({
+})
+
+const store = createStore(reducers, composeEnhancers(applyMiddleware(...middelewares)));
```
Comment (k41n) [inline src/store/store.js:15]: Зачем тут вообще нужен Redux?

**Review by k41n — CHANGES_REQUESTED** [review]
Ok, let's fix more, please have a look at my comments

File: src/components/Login/LoginForm.jsx [hunk 35]
```diff
@@ -0,0 +1,48 @@
...
+
+const LoginForm = (props) => {
+  const { t } = useTranslation();
+  return (
+    <Form
+      onSubmit={props.handleSubmit}
+      render={({ submitError, modifiedSinceLastSubmit, handleSubmit }) => (
+        <div className="flex justify-center mr-8">
+          <form onSubmit={handleSubmit} className="w-full max-w-sm">
+            <EmailField
```
Comment (k41n) [inline src/components/Login/LoginForm.jsx]: We can do even better:
1) translation namespace can be specified only once. `EmailField` component can do `${namespace}.${name}Placeholder` and `${namespace}.${name}` itself.
2) Email validation always applied to the EmailField so component can do that inside

So real props list is:
- namespace
- name
- required (boolean)

File: src/components/Login/LoginForm.jsx [hunk 36]
```diff
@@ -0,0 +1,48 @@
...
+              name="email"
+              placeholder={t("description.emailPlaceholder")}
+              labelText={t("description.email")}
+              validate={composeValidators(required, emailRegex)}
+            />
+            <PasswordField
+              name="password"
+              placeholder={t("description.passwordPlaceholder")}
+              labelText={t("description.password")}
+              validate={composeValidators(required)}
```
Comment (k41n) [inline src/components/Login/LoginForm.jsx]: Component can do that itself, just tell it if it is required. So you don't have to import `composeValidators` and `required` in parent form. It's not its business.

**Review by k41n — CHANGES_REQUESTED** [review]
Almost ideal, some tiny glitches

**Review by k41n — APPROVED** [review]
👍

**Pull request merged**